adcc/LazyMp.py

```diff
@@ -60,39 +57,80 @@ def __getattr__(self, attr):
 
     def df(self, space):
         """Delta Fock matrix"""
-        # TODO This is so cheap to compute that it does not really deserve
-        #      any caching
-        return self.__cpp.df(space)
+        hf = self.reference_state
+        s1 = space[:2]
+        s2 = space[2:]
+        fC = hf.fock(s1 + s1).diagonal()
+        fv = hf.fock(s2 + s2).diagonal()
+        return direct_sum("-i+a->ia", fC, fv).evaluate()
 
+    @cached_member_function
     def t2(self, space):
         """T2 amplitudes"""
-        return self.__cpp.t2(space)
-
-    def set_t2(self, space, tensor):
-        """
-        Set the T2 amplitudes tensor. This invalidates all data depending on the T2
-        amplitudes in this class. Note, that potential other caches, such as
-        computed ADC intermediates are *not* automatically invalidated.
-        """
-        return self.__cpp.set_t2(space, tensor)
+        hf = self.reference_state
+        if space == b.oovv:
```

Comment:
If you have the split-space function from above all these cases can probably unified into one code path. (where you just check that the last two spaces are `v` and then build the df spaces from the first two.

Reply:
Then you can also scratch the compute_t2.

Reply:
Done.